File: CPP/ByRound/1500/1500_C_Matrix_Sorting.cpp
// Problem: CF 1500 C - Matrix Sorting
// https://codeforces.com/contest/1500/problem/C

/*
 * Code Purpose: 
 *   Determines if table A can be transformed into table B using a series of stable column sorts.
 *   If possible, outputs a sequence of columns to sort by.
 *
 * Algorithm:
 *   - Construct a topological order of columns based on the required relative ordering in B.
 *   - Use a greedy method to process columns in topological order.
 *   - Verify that the transformation is valid by reordering rows according to the column sort sequence.
 *
 * Time Complexity: O(H * W^2 + H^2 * W)
 *   - Building dependency graph: O(H * W)
 *   - Topological sort (BFS): O(W + H * W) = O(H * W)
 *   - Checking transformation correctness: O(H^2 * W)
 * 
 * Space Complexity: O(H * W)
 *   - Storage for matrices A, B, and auxiliary arrays.
 */

#include <stdio.h>
#include <algorithm>
#include <queue>
#include <utility>
#include <vector>

using namespace std;
using ll = long long;
using ull = unsigned long long;
#define rep(i, n) for (int i = 0; i < (n); i++)
int H, W;
vector<vector<int>> A, B;
vector<int> F_Row;        // Tracks which rows have not been processed yet
vector<int> C_Col;        // Counts dependencies for each column
vector<int> Ord;          // Stores the sorted column order
int main() {
  scanf("%d%d", &H, &W);
  A.assign(H, vector<int>(W));
  rep(y, H) rep(x, W) scanf("%d", &A[y][x]);
  B.assign(H, vector<int>(W));
  rep(y, H) rep(x, W) scanf("%d", &B[y][x]);
  
  F_Row.assign(H - 1, 1);   // Initially all row pairs are considered unprocessed
  C_Col.assign(W, 0);       // Initialize dependency count for each column
  rep(y, H - 1) rep(x, W) if (B[y][x] > B[y + 1][x]) C_Col[x]++; // Count needed inversions per column
  
  queue<int> Q;
  rep(x, W) if (C_Col[x] == 0) Q.push(x); // Start with columns that have no dependencies
    
  while (Q.size()) {
    int px = Q.front();
    Q.pop();
    Ord.push_back(px);  // Add to the ordered list of columns
    
    rep(y, H - 1) if (B[y][px] < B[y + 1][px]) // Process rows that are incorrectly ordered in this column
      if (F_Row[y]) {
        rep(x, W) if (B[y][x] > B[y + 1][x]) // Reduce dependency count for columns where values are out of order
          if ((--C_Col[x]) == 0) Q.push(x);
        F_Row[y] = 0; // Mark this row pair as processed
      }
  }
  
  reverse(Ord.begin(), Ord.end());  // Reverse to get final valid sorting order
  
  vector<pair<pair<int, int>, int>> P(H);  // Store row index and its current mapping
  rep(y, H) P[y].second = y;
  
  // Apply column sorts in the computed order
  for (int x : Ord) {
    rep(i, H) P[i].first = {A[P[i].second][x], i};  // Build key for sort (value, original index)
    sort(P.begin(), P.end());  // Stable sort by column value
  }
  
  // Check if the reordered rows match matrix B
  rep(i, H) rep(x, W) {
    int y = P[i].second;
    if (A[y][x] != B[i][x]) {
      printf("-1\n");
      return 0;
    }
  }
  
  printf("%d\n", (int)Ord.size());
  rep(i, Ord.size()) {
    if (i)
      printf(" ");
    printf("%d", Ord[i] + 1);  // Output column indices (1-based)
  }
  printf("\n");
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1500/1500_C_Matrix_Sorting.py
# Problem: CF 1500 C - Matrix Sorting
# https://codeforces.com/contest/1500/problem/C

"""
Matrix Sorting Problem

Algorithm:
- This problem involves transforming matrix A into matrix B using sorting operations on columns.
- We model this as a topological sorting problem on a directed graph where nodes represent columns.
- The key insight is to determine if the rows of B can be achieved by sorting columns of A in some order.
- We use a topological sort approach:
  1. For each column, we create a dependency graph based on the relative order of elements in B.
  2. We build a graph where an edge from column i to column j means that sorting on column i must happen before column j.
  3. Perform topological sort to determine the order of columns to sort.
  4. Apply these sorts to A and check if we get B.

Time Complexity: O(n * m^2 + n^2 * m) in the worst case
Space Complexity: O(n * m)

"""

n, m = [int(t) for t in input().split()]
a = [[int(t) for t in input().split()] for _ in range(n)]
b = [[int(t) for t in input().split()] for _ in range(n)]

# Initialize data structures for graph building
inc = []  # inc[i] stores whether column i has any dependency (order constraint)
brick = []  # brick[i] stores row indices where column i has decreasing values

# For each column, track order constraints and decreasing positions
for x in range(m):
    z = []  # List of 1s and 0s indicating if next element is smaller
    w = []  # List of row indices where element decreases
    for y in range(n - 1):
        # If next value is smaller, mark a constraint
        if b[y + 1][x] < b[y][x]:
            z.append(1)
        else:
            z.append(0)
        # If elements differ, note row index for later processing
        if b[y + 1][x] ^ b[y][x]:  # XOR, true if values are different
            w.append(y)
    inc.append(z)
    brick.append(w)

# Calculate out-degree of each column
S = [sum(r) for r in inc]  # S[i] counts how many dependencies column i has
v = [0] * n  # visited array for tracking row processing

col = []  # Store the topologically sorted order of columns
stack = [i for i in range(m) if S[i] == 0]  # Start with columns having no dependencies

# Perform topological sort
while stack:
    g = stack.pop()  # Get column with no more dependencies
    col.append(g)  # Add to result order
    # For each row where column g has a change
    for e in brick[g]:
        if v[e]:  # Skip if already processed
            continue
        v[e] = 1  # Mark row as processed
        # Check all columns for dependencies on this row
        for x in range(m):
            if inc[x][e]:  # If column x depends on row e
                S[x] -= 1  # Reduce dependency count
                if S[x] == 0:  # If no more dependencies
                    stack.append(x)  # Add to processing queue

# Apply the column sorts in reverse order (topological sort)
for e in reversed(col):
    a.sort(key=lambda x: x[e])

# Check if transformation succeeded
if a == b:
    print(len(col))
    print(*[e + 1 for e in reversed(col)])  # Output 1-indexed column numbers
else:
    print(-1)  # Transformation impossible


# https://github.com/VaHiX/CodeForces/